specification/k3po.control/SPEC.md

```diff
@@ -0,0 +1,258 @@
+
+## Requirements
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "MAY",
+and "OPTIONAL" in this document are to be interpreted as described in [RFC2119](https://tools.ietf.org/html/rfc2119).
+
+## The Protocol
+
+The protocol is made up of Commands and Events.  Commands are actions that can be triggered by the test framework.
+Events are triggered by observed actions or state changes in the driver and are sent back to the test framework.  The protocol 
+is formatted by:
+
+>Control Message   =   command-line LF
+>                      *( header-field LF )
+>                      LF
+>                      [ message-body LF ]
+
+command-line      = command LF
+
+command              = token
+
+header-field = field-name ":" field-value
+
+LF, message-body, field-name, field-value, and token are defined in [HTTP RFC](https://tools.ietf.org/html/rfc7230).
+
+If the "content-length" header is specified, then a message body of the specified length must be added.
+
+Headers that are not understood / processed by a command must be ignored.
+
+
+### Flow
+
+The communication between the test framework and the driver is a conversation - the test framework sends a command, the driver responds with an
+event sent asynchronously when the action is finished. The only exception is the FINISHED event - the START command will have two events as a 
+response - STARTED when the execution of the scripts is started and FINISHED when the execution of the scripts is finished.
+
+Throughout the flow description bellow, **an error** represents an ERROR event sent by the driver.
+
+- The test framework initiates the conversation with a **PREPARE** command. This command will make the k3po server load and prepare a set
+of scripts for execution. The scripts are not yet executed, just loaded from disk and parsed. If there are ACCEPT instructions in the scripts,
+a bound on corresponding ports is performed. If the command is successful, the driver sends back a PREPARED event. In case of failure, an
+ERROR event is sent.
+
+- If the PREPARED command is successful, a **START** command may be sent by the test framework. This command will start the execution of the
+ clients in the scripts that were sent in the PREPARE command. If the command is successful, the driver sends back a STARTED event. This 
+ command will also trigger the sending of a FINISHED event whenever the scripts finished execution on the k3po server.
+ 
+- If there is no PREPARE command send before the START command or if the PREPARE command was not successful an error with a message "Script has not
+ been prepared or is still preparing" must occur. 
+ 
+- If the START command was already sent, an error with a message "Script has already been started" must occur.
+
+- Any time after the PREPARE command and before the DISPOSE command, the test framework can send an **AWAIT** command for a barrier existing 
```

Comment:
I think it is important to note that The NOTIFIED event may never be sent back if that barrier is never triggered, so the test framework can't block on it